src/compiler/checker.ts

```diff
@@ -24868,6 +24869,18 @@ namespace ts {
 
             if (isStatic(container) || isCallExpression) {
                 nodeCheckFlag = NodeCheckFlags.SuperStatic;
+                if (!isCallExpression &&
+                    languageVersion >= ScriptTarget.ES2015 && languageVersion <= ScriptTarget.ES2021 &&
+                    (isPropertyDeclaration(container) || isClassStaticBlockDeclaration(container))) {
+                    // for `super.x` or `super[x]` in a static initializer, mark all enclosing
+                    // block scope containers so that we can report potential collisions with
+                    // `Reflect`.
+                    forEachEnclosingBlockScopeContainer(node.parent, current => {
+                        if (!isSourceFile(current) || isExternalOrCommonJsModule(current)) {
+                            getNodeLinks(current).flags |= NodeCheckFlags.ContainsSuperPropertyInStaticInitializer;

```

Comment:
how does the emitter know to emit the `Reflect` class? Or is it unconditionally emitted, and the author just needs to know to provide one?

Reply:
`Reflect` is a global in ES2015. We base this on the minimum target in the same way that we use `WeakMap` and `WeakSet` for private fields and `Promise` for async functions. It isn't an issue in ES5/3 since we use `_super` and don't support regular property access on `super`, only method invocation.

Reply:
Oh, so the reason it's a problem is that a local `Reflect` will shadow the global `Reflect` which our emit assumes is available. That makes sense.